### DISCUSSION TITLE
Differential ADC vs Single Ended ADC

### DISCUSSION THREAD

**Discussion by darrenweinhold** [micropython/micropython#9906] [General]
Is there a way to configure ADC channels as differential instead of single ended? I can't find any mention of it in the documentation and don't see anything in the source code for it.

**Comment by davefes** [comment]
Which bit of hardware?

**Reply by darrenweinhold** [reply]
I'm looking specifically at the Portenta H7. But I was thinking more generally that there would likely be a way to do this for hardware that supports it.

**Reply by rkompass** [reply]
The processor of Portenta H7, the STM32H747 as an ADC that provides differential inputs:  _Can manage Single-ended or differential inputs (programmable per channels)._ This is stated on page 962 of the [reference manual](https://www.st.com/resource/en/reference_manual/rm0399-stm32h745755-and-stm32h747757-advanced-armbased-32bit-mcus-stmicroelectronics.pdf). The assignment of (the 20 relevant) pins to positive and negative input channels is fixed and can be seen at page 968. Apparently you just have to set a bit in the **ADC differential mode selection register (ADC_DIFSEL)** to activate the differential mode for a channel (should be done with inactive ADC), see page 1075.
This you may achieve with a mem32() instruction from micropython. Address offset: 0xC0, didn't find the base address yet.
It is unlikely that such a feature will be incorporated into MP soon, as there are lots of other features in different processors that would have to be coordinated before, to have a common api. But you might write an adc class especially for this processor.

**Reply by darrenweinhold** [reply]
Thanks for the feedback. I missed the note on page 1075. I will do some more digging to see about setting the register. It would be great to have some defined way of using differential mode, but I understand the challenges with some much different hardware to support. 

I really appreciate the thoroughness of this answer and how great answers like this are to the community. Thanks!

**Comment by rkompass** [comment]
Not on the Pyboards, not on RPi2040 (the hardware does not allow this). There are external ADCs with drivers for micropython like for example the [ads1115](https://github.com/robert-hh/ads1x15/blob/master/ads1x15.py) which allow to configure differential inputs.

**Reply by darrenweinhold** [reply]
That is very good to know for that hardware. I am looking at an ADS1115 as a backup :)
